File: frontend/EduMatch/src/types/application.ts
export type ApplicationStatus = "pending" | "accepted" | "rejected";

export interface Application {
  id: number;
  request_id: number;
  instructor_id: number;
  message: string;
  proposed_price: string;
  status: ApplicationStatus;
  created_at: string;
  updated_at: string;
  instructor_name: string | null;
  instructor_specialization: string | null;
  instructor_rating: string | null;
  request_title: string | null;
  student_name: string | null;
}

export interface ApplicationCreatePayload {
  request_id: number;
  message: string;
  proposed_price: number;
}

export interface ApplicationDecisionResponse {
  application: Application;
  session: import("./session").Session | null;
}

File: backend/app/models/application.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Application(Base):
    __tablename__ = "applications"
    __table_args__ = (UniqueConstraint("request_id", "instructor_id", name="uq_application_request_instructor"),)

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    request_id: Mapped[int] = mapped_column(ForeignKey("requests.id", ondelete="CASCADE"), index=True, nullable=False)
    instructor_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    proposed_price: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="pending", server_default="pending")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    request = relationship("LearningRequest", back_populates="applications")
    instructor = relationship("User", back_populates="applications")
    messages = relationship("Message", back_populates="application", cascade="all, delete-orphan")
